photon-api/src/main/scala/com/linkedin/photon/ml/algorithm/FixedEffectCoordinate.scala

```diff
@@ -72,83 +61,100 @@ protected[ml] class FixedEffectCoordinate[Objective <: DistributedObjectiveFunct
+    dataset: DataFrame,
+    optimizationProblem: DistributedOptimizationProblem[Function],
+    featureShardId: FeatureShardId,
+    inputColumnsNames: InputColumnsNames,
+    initialFixedEffectModelOpt: Option[FixedEffectModel]): (FixedEffectModel, OptimizationTracker) = {
+
+    val rdd = dataset
+      .rdd
+      .map { row =>
+        val uid = row.getAs[UniqueSampleId](DataConst.ID)
+        val features = row.getAs[SparkVector](featureShardId)
+        val label = row.getAs[Double](inputColumnsNames(InputColumnsNames.RESPONSE))
+
+        (uid, LabeledPoint(label, VectorUtils.mlToBreeze(features)))
       }
-      .getOrElse(optimizationProblem.runWithSampling(input))
-    val updatedModelBroadcast = input.sparkContext.broadcast(newModel)
+    rdd.persist(StorageLevel.MEMORY_ONLY)
 
-    new FixedEffectModel(updatedModelBroadcast, featureShardId)
-  }
+    val (glm, stateTracker) = initialFixedEffectModelOpt
+      .map ( initialFixedEffectModel =>
+        optimizationProblem.runWithSampling(rdd, initialFixedEffectModel.model)
+      )
+      .getOrElse(optimizationProblem.runWithSampling(rdd))
 
-  /**
-   * Score a [[FixedEffectDataset]] using a given [[FixedEffectModel]].
-   *
-   * @note The score is the dot product of the model coefficients with the feature values (i.e., it does not go
-   *       through a non-linear link function).
-   * @param fixedEffectDataset The dataset to score
-   * @param fixedEffectModel The model used to score the dataset
-   * @return The computed scores
-   */
-  protected[algorithm] def score(
-      fixedEffectDataset: FixedEffectDataset,
-      fixedEffectModel: FixedEffectModel): CoordinateDataScores = {
+    rdd.unpersist()
 
-    val modelBroadcast = fixedEffectModel.modelBroadcast
-    val scores = fixedEffectDataset.labeledPoints.mapValues { case LabeledPoint(_, features, _, _) =>
-      modelBroadcast.value.computeScore(features)
-    }
+    (FixedEffectModel(glm, featureShardId), new FixedEffectOptimizationTracker(stateTracker))
+  }
 
-    new CoordinateDataScores(scores)
+  def updateOffset(
+    dataset: DataFrame, fixedEffectModel: FixedEffectModel, featureShardId: FeatureShardId,
+    inputColumnsNames: InputColumnsNames): DataFrame = {
+
+    require(
+      featureShardId == fixedEffectModel.featureShardId,
+      s"Fixed effect coordinate featureShardId ${featureShardId} != model.featureShardId ${
+        fixedEffectModel
+          .featureShardId
+      }")
+
+    val offset = inputColumnsNames(InputColumnsNames.OFFSET)
+    val hasOffsetField = ApiUtils.hasColumn(dataset, offset)
+    val hasCoordinateScoreField = ApiUtils.hasColumn(dataset, SCORE_FIELD)
+
+    if (hasOffsetField && hasCoordinateScoreField) {
+      // offset = offset - old_coordinateScore + new_coordinateScore
+      dataset.withColumn(offset, col(offset) - col(SCORE_FIELD))
+      fixedEffectModel.computeScore(dataset, SCORE_FIELD)
```

Comment:
where is the new score saved?

Reply:
I think new score is saved in SCORE_FIELD.